### ISSUE TITLE
TTF generated with only bitmaps doesn't render in Windows

### ISSUE THREAD

**Issue by sunaku** [fontforge/fontforge#2738]
- [x] Screenshot ![](http://i.imgur.com/N8F9F6Q.png)
- [x] The FontForge version and the operating system you're using
  - To view the generated TTF: Windows 7
  - To generate the TTF:
```sh
$ fontforge --version
Copyright (c) 2000-2014 by George Williams. See AUTHORS for Contributors.
 License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
 with many parts BSD <http://fontforge.org/license.html>. Please read LICENSE.
 Based on sources from 15:13 UTC  5-Apr-2016-ML-D.
 Based on source from git with hash:
no xdefs_filename!
TESTING: getPixmapDir:/usr/share/fontforge/pixmaps
TESTING: getShareDir:/usr/share/fontforge
TESTING: GResourceProgramDir:/usr/sbin
trying default theme:/usr/share/fontforge/pixmaps/resources
fontforge 15:13 UTC  5-Apr-2016
libfontforge 20160405

$ uname -a
Linux yantram 4.6.3_1 #1 SMP PREEMPT Sat Jun 25 05:40:29 UTC 2016 x86_64 GNU/Linux
```
- [x] The behavior you expect to see, and the actual behavior
  - Expected: Bitmaps inside the TTF should appear in Windows Font Viewer.
  - Actual: Bitmaps inside the TTF **do not** appear in Windows Font Viewer.
- [x] Steps to reproduce the behavior 
  1. Download https://github.com/sunaku/tamzen-font/archive/bdf2ttf-fontforge.zip
  2. Extract it and go inside the extracted folder.
  3. (optional) Run `fontforge -script ttf/bdf2ttf.pe ttf/Tamzen-Regular.ttf bdf/Tamzen[0-9]*r.bdf` to generate the output `ttf/Tamzen-Regular.ttf` file.
  5. (optional) Open the `ttf/Tamzen-Regular.sfd` file and observe that bitmaps are actually present.
  6. (optional) Open the `ttf/Tamzen-Regular.ttf` file and observe that bitmaps are actually present.
![1467660763](https://cloud.githubusercontent.com/assets/9863/16569098/2048034c-41e7-11e6-8221-f2240b62d29b.png)
  7. Double-click on the output `ttf/Tamzen-Regular.ttf` file under Windows and observe that none of the bitmaps are rendered (as shown in the main screenshot at the top of this bug report).
- [ ] \(optional) Possible solution/fix/workaround

**Comment by ctrlcctrlv** [comment]
>Expected: Bitmaps inside the TTF should appear in Windows Font Viewer.
Do you have a TTF made of only bitmaps that does appear?

**Comment by sunaku** [comment]
Yes, please see [these TTFs](https://github.com/koron/bdf2ttf/files/344821/TamzenForPowerline.zip) generated from the same BDFs in https://github.com/koron/bdf2ttf/issues/1#issuecomment-230130135.

**Comment by ctrlcctrlv** [comment]
Sorry, I'm using Windows 10 and can't get a preview on that archive with fontview.exe, but viewing it in FontForge seems okay.

![2016-07-06 8](https://cloud.githubusercontent.com/assets/838783/16613130/3c2a963a-439d-11e6-9d47-3c0976e66898.png)

The reason that I'm asking this of you is so that I can examine both in TTX and see what error FontForge is making.

I'm also somewhat dubious over whether Windows even supports these kinds of TTF files.

**Comment by sunaku** [comment]
Here's the result of double-clicking the TTFs (in the ZIP file [I linked to earlier](https://github.com/fontforge/fontforge/issues/2738#issuecomment-230664832)) under Windows 7:

![screenshot](http://i.imgur.com/zo9zNTY.png)

Indeed, these fonts _**do** work properly_ under Windows 7, as demonstrated by Notepad here: :trophy: 

![screenshot](http://i.imgur.com/bjwGixC.png)

**Comment by flowchartsman** [comment]
Over a year with a fairly serious issue and no progress or updates? What's the story, guys?

**Comment by khaledhosny** [comment]
Fix it yourself?

**Comment by flowchartsman** [comment]
Still broken. Coming up on the two year anniversary.

**Comment by ctrlcctrlv** [comment]
@flowchartsman This is a community maintained project. No one is paid to work on this. I also would definitely disagree that this issue is in any way severe. A TrueType font containing only bitmaps is a rarity in the first place, and their utility in any situation over dedicated pixel font formats is incredibly dubious.

Your comments sound entitled. I think you should change your perspective.

**Comment by flowchartsman** [comment]
I have returned to this issue from some time away and first I would like to publicly apologize for my impatience before. My tone was inappropriate.

Second, I would like to offer a couple paths forward for anyone stuck in a similar situation. If you are keen on using bitmap fonts simulated with outlines, the [Bits'N'Picas](https://github.com/kreativekorp/bitsnpicas) project might be a good alternative, since it is explicitly designed for this usecase. I have used it with great success, however it is not as generally flexible or powerful as FontForge, nor does it support bitmap strikes. This approach is not without its caveats, since apparently ClearType must be disabled, which must be done system-wide. This might be able to be bypassed if the application itself explicitly supports disabling aliasing (such as VSCode), but I don't have a Windows Machine to test.

For getting bitmap strikes to work, my understanding is that Windows has explicitly disabled these in fonts, with one exception: fonts with CKJ codepages.  From reading up on the issue it seems this is still allowed because the complexity of characters in these codepages are often such that bitmap strikes are the only way to ensure that they are visually distinct enough for identification onscreen at certain sizes. (more [here](https://int10h.org/blog/2016/01/windows-cleartype-truetype-fonts-embedded-bitmaps/)), and the requirements for getting them back are:
> 1. At least one CJK code page must be specified in the font's 'OS/2' table (e.g.  Chinese Traditional, or JIS/Japan); and
> 2. A few seemingly-random East Asian glyphs actually have to exist in the font.

However, this apparently comes with some caveats, namely that the font can no longer be used in certain circumstances, namely in console windows :(

So it seems as if emulation with outlines is really the only surefire method to get pixel-perfect font rendering, if that's what you want.

**Comment by ctrlcctrlv** [comment]
@flowchartsman you are not the only one with new ideas since 2016. :-)

Please try [bitmapfont2otb](https://github.com/ctrlcctrlv/bitmapfont2otb) and let me know if it works for you…it uses FF API.
